src/main/java/org/acme/consumer/ConfigConsumer.java

```diff
@@ -0,0 +1,84 @@
+package org.acme.consumer;
+
+import alpine.model.ConfigProperty;
+import io.quarkus.runtime.StartupEvent;
+import org.acme.serde.ConfigPropertyDeserializer;
+import org.acme.serde.ConfigPropertySerializer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.StoreQueryParameters;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.errors.InvalidStateStoreException;
+import org.apache.kafka.streams.kstream.GlobalKTable;
+import org.apache.kafka.streams.kstream.Materialized;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.QueryableStoreTypes;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import java.util.Properties;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Objects;
+
+@ApplicationScoped
+public class ConfigConsumer {
```

Comment:
For this one we had a slightly modified idea around what you proposed. 
Let’s have a notification topic that sends a notification to our stateless quarkus app whenever there is a change in DT config. If there is a change in config then this notification topic will have an event. We want this event to be listened to by one consumer. When an event occurs, it gets the latest config from DT and updates the info in the in memory structure of quarkus apps. This way we would not be storing global config as such in ktable (like suggested yesterday) and we would only be fetching latest config from dt when there is a change (otherwise use existing values) — so there is no unnecessary polling. I think it is a good idea. What do you think?

Reply:
Implemented this using HashMap instead of Global K Table

Reply:
> For this one we had a slightly modified idea around what you proposed. Let’s have a notification topic that sends a notification to our stateless quarkus app whenever there is a change in DT config. If there is a change in config then this notification topic will have an event. We want this event to be listened to by one consumer. When an event occurs, it gets the latest config from DT and updates the info in the in memory structure of quarkus apps. This way we would not be storing global config as such in ktable (like suggested yesterday) and we would only be fetching latest config from dt when there is a change (otherwise use existing values) — so there is no unnecessary polling. I think it is a good idea. What do you think?

That's an interesting alternative. There would be three things to consider here 

1. Multi-instance target : All of the instances of the quarkus stateless app should receive the event to update their in-memory state. Right now, since you are using the same `common.consumer-config-app-id` which translate to the kafka consumer `group.id`, the events will be partioned and partially processed by your cluster of quarkus app because of Kafka consumer group protocol. You need to use a random application identifier like `common.consumer-config-app-id` + `generated UUID` to have each instance receive all of the events
2. Initial state : How do you build the in-memory state at startup ? That is a challenge because even though you can manually send event for your tests, you have to take into account, instance restart or reschedule. There's a partial solution with the random application identifier I suggested above because using a combination of random `APPLICATION_ID` and `AUTO_OFFSET_RESEST=earliest` will make the consumer consume all of the events available in the topic to reconstruct it's in-memory state. That's a partial solution because you also have to take into account **Topic purge** and **Topic volumetry**. To completely solve this part, you need to make the config topic [a log compacted one](https://kafka.apache.org/documentation/#compaction)
3.  Kafka Streams usage : Kafka streams use case is read/process/write from/to Kafka. If you only need to read from Kafka to build in-memory state, there is no need for a Kafka Streams topology

I'm still concerned by this design because, at a given moment, due to consumption lag or network issues, different components from the same BOM can be processed by different instances of quarkus stateless app with different in-memory states (`snyk.enable` being simultaneously `true` and `false`). Debugging issues like that will prove very difficult

Is it really that much of an issue to read properties from DT ApiServer REST API ? At the very least, It should be benched.

Reply:
Thanks for your input Alioune..I will take into consideration the suggestions provided by you..Regarding fetching config from Dt every time, we think we would bombard DT with too many post requests as we will then have to make request for each component, as DT is sending individual component request to quarkus service

Reply:
I also think it's probably not as big of a problem to call the API server often, assuming some sort of caching is in place. 

> At the very least, It should be benched.

That is the key takeaway here I assume. Should be trivial to setup using something like https://github.com/rakyll/hey